File: clojure/009/009.clj
(defn pythagorean_triple_with_sum
  ([sum]
    (loop [m 2 n 1]
      (if (and (and (< m sum) (< n (- sum 1))) (> m n))
        (let [a (- (* m m) (* n n)) b (* 2 m n) c (+ (* m m) (* n n))]
          (if (and (= (+ a b c) sum) (= (+ (* a a) (* b b)) (* c c)))
            (list a b c)
            (recur m (+ n 1))
          )
        )
        (if (and (< m sum) (< n (- sum 1)))
          (recur (+ 1 m) 1)
          '()
        )
      )
    )
  )
)

(println (apply * (pythagorean_triple_with_sum 1000)))

File: ruby/009/009.rb
def pythagorean_triple_with_sum(sum)
  raise ArgumentError, 'The smallest possible sum of a pythagorean triple is 11.' if sum < 11

  (2..sum).each do |m|
    (1..(sum - 1)).each do |n|
      next unless m > n

      a = (m * m) - (n * n)

      b = 2 * m * n

      c = (m * m) + (n * n)

      return [a, b, c] if (a + b + c == sum) && ((a * a) + (b * b) == (c * c))
    end
  end

  []
end

puts pythagorean_triple_with_sum(1000).inject(:*)
